### PR TITLE
V1 rewrite

### LINKED ISSUE #4: Consistent error behavior in Configure()
As of 2018-06-25, the `Configure` method issues a log message for some _but not all_ errors. It ought to have more consistent behavior. My inclination is to log for everything, since some of the fixed `Err...` returns don't have all the information you might reasonably want, but I'm open to other opinions.

### LINKED ISSUE #25: Add error return to PKCS11Session.Close (breaking change)
During the work for #24, it became clear that `PKCS11Session.Close()` ought to return an error but doesn't.

This would be a breaking change, so the interim fix was to add an alternative `PKCS11Session.CloseSession()` method that does return the error. The `PKCS11Session.Close()` was deprecated.

During the next breaking change release, we should either remove the `Close()` method or change its signature.

### LINKED ISSUE #29: Return error when Configure or ConfigureFromFile has not been called
If a user forgets to call `Configure` or `ConfigureFromFile`, panics and segfaults may ensue due to using an uninitialised copy of `miekg/pkcs11`. Those are the only two functions that call `pkcs11.New`.

We should catch this situation earlier and return descriptive errors.

### LINKED ISSUE #30: Add information about Configure and ConfigureFromFile to README
The README contains no discussion of calling `Configure` or `ConfigureFromFile`. We should change this so that new users have an easier path to working code.

### LINKED ISSUE #35: Panic due to misaligned atomic variables
In the dependency vitessio/vitess, a panic is caused on certain architectures due to misaligned variables. This will be fixed by vitessio/vitess#4729. When this is merged, the dependency will need to be re-vendored.

### LINKED ISSUE #37: Copyright missing in the common.go file
Copyright is missing in crypto11/common.go file. All the other go files have it.

### PR DESCRIPTION
**Pull request by dmjones**
Re-write of crypto11 to implement various backwards-incompatible design changes. This PR will be released as v1.0.0, beginning a new commitment to API stability.

Many of the topics in #36 are addressed, plus several outstanding issues:

- Singleton pattern is no more, this closes #29 (because it's no longer relevant)
- No exposure of underlying PKCS#11 connection information, ensuring our thread safety commitments are maintained.This closes #25 (because it's no longer relevant).
- No more random IDs and labels. You can either specify an ID, or an ID and label.
- Removal of slot and session-based key generation.
- Missing copyright added, closes #37 
- Vitess dependency has been updated, closes #35 
- Improved documentation, closes #30 
- Remove logging (which doesn't belong in a package, IMO). This closes #4.
- Remove named errors from our exported API.
- Attempts to find missing keys now returns nil, not a named error.

The main thing discussed in #36 that wasn't implement was the `WithAttributes` functions. I'm not sure they are needed, so I'm happy to wait for someone to request them. I'll create an issue to capture that as a potential enhancement. There is a branch, [`with-attributes`](https://github.com/ThalesIgnite/crypto11/tree/with-attributes), the contains the beginning of some work on this. #20 also hasn't been implemented, but it will be.

### CODE REVIEW

File: ecdsa.go [hunk 1]
```diff
@@ -162,48 +162,35 @@ func unmarshalEcParams(b []byte) (elliptic.Curve, error) {
...
 
-func unmarshalEcPoint(b []byte, c elliptic.Curve) (x *big.Int, y *big.Int, err error) {
-	// Decoding an octet string in isolation seems to be too hard
-	// with encoding.asn1, so we do it manually. Look away now.
-	if b[0] != 4 {
+func unmarshalEcPoint(b []byte, c elliptic.Curve) (*big.Int, *big.Int, error) {
+	var pointBytes []byte
+	extra, err := asn1.Unmarshal(b, &pointBytes)
+	if err != nil {
 		return nil, nil, ErrMalformedDER
```
Comment (nickrmc83) [inline ecdsa.go]: Given we now have an error from asn1 Unmarshal we should return that.

**Review by nickrmc83 — CHANGES_REQUESTED** [review]
Just a single small change from a quick scan.

File: ecdsa.go (hunk 1 above)
Comment (dmjones) [reply ecdsa.go]: In line with our conversations yesterday, there are no exported errors any more. This particular bit of code now returns the underlying error (with decoration).

**Comment by dmjones** [comment]
I've pushed a few more changes. Mostly about removing errors from our API. However, I decided to adjust the find keys stuff so it returns nil if the keys aren't found, not an error.

**Review by nickrmc83 — APPROVED** [review]
LGTM

**Comment by cbroglie** [comment]
The tag for this release is `v.1.0.0`, which I'm guessing was a typo (note extra dot: `v.`)

**Comment by dmjones** [comment]
Thanks @cbroglie, I've just corrected that.

**Pull request merged**